src/common/libhostlist/hostlist.c

```diff
@@ -0,0 +1,1140 @@
+    if (!orig)
+        return -1;
+
+    if ((p = strchr (str, '-'))) {
+        *p++ = '\0';
+        if (*p == '-') {  /* don't allow negative numbers */
+            errno = EINVAL;
+            goto error;
+        }
+    }
+    range->lo = strtoul (str, &q, 10);
+    if (q == str)  {
+        errno = EINVAL;
+        goto error;
+    }
+
+    range->hi = (p && *p) ? strtoul (p, &q, 10) : range->lo;
+
+    errno = EINVAL;
+    if (q == p || *q != '\0')
+        goto error;
+
+    if (range->lo > range->hi)
+        goto error;
+
+    if (range->hi - range->lo + 1 > MAX_RANGE ) {
+        errno = ERANGE;
+        goto error;
+    }
+
+    free (orig);
+    range->width = strlen (str);
+    return 0;
+
+  error:
+    free(orig);
+    return -1;
+}
+
+static int hostlist_append_host (struct hostlist *hl, const char *str)
+{
+    struct hostrange * hr;
+    struct hostname * hn;
+
+    assert (hl != NULL);
+
+    if (str == NULL || *str == '\0')
+        return 0;
+
+    if (!(hn = hostname_create (str)))
+        return -1;
+
+    if (hostname_suffix_is_valid (hn)) {
+        hr = hostrange_create (hn->prefix,
+                               hn->num,
+                               hn->num,
+                               hostname_suffix_width(hn));
+    } else
+        hr = hostrange_create_single (str);
+
+    hostlist_append_range (hl, hr);
+
+    hostrange_destroy(hr);
+    hostname_destroy(hn);
+
+    return 1;
+}
+
+/*
+ * Convert 'str' containing comma separated digits and ranges into an array
+ *  of struct _range types (max 'len' elements).
+ *
+ * Return number of ranges created, or -1 on error.
+ */
+static int parse_range_list (char *str, struct _range *ranges, int len)
+{
+    char *p;
+    int count = 0;
+
+    while (str) {
+        if (count == len)
+            return -1;
+        if ((p = strchr (str, ',')))
+            *p++ = '\0';
+        if (parse_next_range (str, &ranges[count++]) < 0)
+            return -1;
+        str = p;
+    }
+    return count;
+}
+
+static void append_range_list (struct hostlist *hl,
+                               char *pfx,
+                               struct _range *rng,
+                               int n)
+{
+    int i;
+    for (i = 0; i < n; i++) {
+        hostlist_append_hr (hl, pfx, rng->lo, rng->hi, rng->width);
+        rng++;
+    }
+}
+
+static void append_range_list_with_suffix (struct hostlist *hl,
+                                           char *pfx,
+                                           char *sfx,
+                                           struct _range *rng,
+                                           int n)
+{
+    int i;
+    unsigned long j;
+    for (i = 0; i < n; i++) {
+        for (j = rng->lo; j <= rng->hi; j++) {
+            char host[4096];
+            struct hostrange * hr;
+            snprintf (host, 4096, "%s%0*lu%s", pfx, rng->width, j, sfx);
+            hr = hostrange_create_single (host);
+            hostlist_append_range (hl, hr);
```

Comment:
check for `hostrange_create_single()` failure
check for `hostliust_append_range()` failure